## Online Retail sales:
This dataset contains all purchases made for an online retail company based in the UK during an eight month period.
We will analyse the monthly sales 

#### Import Libraries & Dataset

In [1]:
from datetime import datetime, timedelta
import pandas as pd
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
from __future__ import division

from chart_studio import plotly
import plotly.offline as pyoff
import plotly.graph_objs as go




In [2]:
df_Retail =pd.read_csv(r"OnlineRetail.csv",encoding='unicode_escape')

df_Retail.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

#### EDA

In [3]:
df_Retail.shape

(541909, 8)

The dataset contains 8 columns and .5 Million records

In [4]:
df_Retail.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [5]:
df_Retail.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

Apart of the two columns, Stock Code and Customer ID, no other column has any missing value

In [6]:
df=df_Retail
df1 = df

#### EDA

In [7]:
df1.Country.nunique()

38

In [8]:
df1.Country.unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

In [9]:
customer_country=df1[['Country','CustomerID']].drop_duplicates()

customer_country.groupby(['Country'])['CustomerID'].aggregate('count').reset_index().sort_values('CustomerID', ascending=False)

Country  CustomerID
36        United Kingdom        3950
14               Germany          95
13                France          87
31                 Spain          31
3                Belgium          25
33           Switzerland          21
27              Portugal          19
19                 Italy          15
12               Finland          12
1                Austria          11
25                Norway          10
24           Netherlands           9
0              Australia           9
6        Channel Islands           9
9                Denmark           9
7                 Cyprus           8
32                Sweden           8
20                 Japan           8
26                Poland           6
34                   USA           4
5                 Canada           4
37           Unspecified           4
18                Israel           4
15                Greece           4
10                  EIRE           3
23                 Malta           2
35  United Arab Emirates           2
2                Bahrain           2
22             Lithuania           1
8         Czech Republic           1
21               Lebanon           1
28                   RSA           1
29          Saudi Arabia           1
30             Singapore           1
17               Iceland           1
4                 Brazil           1
11    European Community           1
16             Hong Kong           0

Majority of data is from UK. So, we will restrict our scope of project to UK only.

In [10]:
df1 = df1.loc[df1['Country'] == 'United Kingdom']

In [11]:

df1.isnull().sum(axis=0)

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     133600
Country             0
dtype: int64

Removing Null Values:

In [12]:
df1 = df1[pd.notnull(df1['CustomerID'])]
df1.isnull().sum(axis=0)

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [13]:
df1.describe()

Quantity      UnitPrice     CustomerID
count  361878.000000  361878.000000  361878.000000
mean       11.077029       3.256007   15547.871368
std       263.129266      70.654731    1594.402590
min    -80995.000000       0.000000   12346.000000
25%         2.000000       1.250000   14194.000000
50%         4.000000       1.950000   15514.000000
75%        12.000000       3.750000   16931.000000
max     80995.000000   38970.000000   18287.000000

Quantity is -ve value in some records. As quantity can't be negative. We have to remove them

In [14]:
df1 = df1[(df1['Quantity']>0)]
df1.Quantity.min()

1

In [15]:
df1.shape

(354345, 8)

In [16]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 354345 entries, 0 to 541893
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    354345 non-null  object 
 1   StockCode    354345 non-null  object 
 2   Description  354345 non-null  object 
 3   Quantity     354345 non-null  int64  
 4   InvoiceDate  354345 non-null  object 
 5   UnitPrice    354345 non-null  float64
 6   CustomerID   354345 non-null  float64
 7   Country      354345 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 24.3+ MB


In [17]:
df1.nunique()

InvoiceNo      16649
StockCode       3645
Description     3844
Quantity         294
InvoiceDate    15615
UnitPrice        403
CustomerID      3921
Country            1
dtype: int64

In [18]:
df1['TotalPrice'] = df1['Quantity'] * df1['UnitPrice']
df1.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  TotalPrice  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom       15.30  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom       22.00  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom       20.34

#### Setting Recency:

In [20]:
df1['InvoiceDate'].min()

'1/10/2011 10:32'

In [21]:
df1['InvoiceDate'].max()

'9/9/2011 9:52'

In [27]:
import datetime as dt
NOW = dt.datetime(2011,12,10)

In [28]:
df1['InvoiceDate'] = pd.to_datetime(df1['InvoiceDate'])

In [29]:
rfmTable = df1.groupby('CustomerID').agg({'InvoiceDate': lambda x: (NOW - x.max()).days, # Recency
                                        'InvoiceNo': lambda x: len(x),      # Frequency
                                        'TotalPrice': lambda x: x.sum()}) # Monetary Value

rfmTable['InvoiceDate'] = rfmTable['InvoiceDate'].astype(int)
rfmTable.rename(columns={'InvoiceDate': 'recency', 
                         'InvoiceNo': 'frequency', 
                         'TotalPrice': 'monetary_value'}, inplace=True)

In [30]:
rfmTable.head()

recency  frequency  monetary_value
CustomerID                                    
12346.0         325          1        77183.60
12747.0           2        103         4196.01
12748.0           0       4596        33719.73
12749.0           3        199         4090.88
12820.0           3         59          942.34

In [31]:
quantiles = rfmTable.quantile(q=[0.25,0.5,0.75])
quantiles

recency  frequency  monetary_value
0.25     17.0       17.0          300.04
0.50     50.0       41.0          651.82
0.75    142.0       99.0         1575.89

In [32]:
quantiles = quantiles.to_dict()
quantiles

{'recency': {0.25: 17.0, 0.5: 50.0, 0.75: 142.0},
 'frequency': {0.25: 17.0, 0.5: 41.0, 0.75: 99.0},
 'monetary_value': {0.25: 300.03999999999996,
  0.5: 651.8199999999999,
  0.75: 1575.89}}

In [33]:
segmented_rfm = rfmTable

In [34]:
def RScore(x,p,d):
    if x <= d[p][0.25]:
        return 1
    elif x <= d[p][0.50]:
        return 2
    elif x <= d[p][0.75]: 
        return 3
    else:
        return 4
    
def FMScore(x,p,d):
    if x <= d[p][0.25]:
        return 4
    elif x <= d[p][0.50]:
        return 3
    elif x <= d[p][0.75]: 
        return 2
    else:
        return 1

In [35]:
segmented_rfm['r_quartile'] = segmented_rfm['recency'].apply(RScore, args=('recency',quantiles,))
segmented_rfm['f_quartile'] = segmented_rfm['frequency'].apply(FMScore, args=('frequency',quantiles,))
segmented_rfm['m_quartile'] = segmented_rfm['monetary_value'].apply(FMScore, args=('monetary_value',quantiles,))

In [36]:
segmented_rfm.head()

recency  frequency  monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
12346.0         325          1        77183.60           4           4   
12747.0           2        103         4196.01           1           1   
12748.0           0       4596        33719.73           1           1   
12749.0           3        199         4090.88           1           1   
12820.0           3         59          942.34           1           2   

            m_quartile  
CustomerID              
12346.0              1  
12747.0              1  
12748.0              1  
12749.0              1  
12820.0              2

The customer 12346 is high spending. So, we can market him our expensive products. The customer 12747 & 12748 are our best customers

In [39]:
segmented_rfm['RFMScore'] = segmented_rfm.r_quartile.map(str) \
                            + segmented_rfm.f_quartile.map(str) \
                            + segmented_rfm.m_quartile.map(str)
segmented_rfm.head()

recency  frequency  monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
12346.0         325          1        77183.60           4           4   
12747.0           2        103         4196.01           1           1   
12748.0           0       4596        33719.73           1           1   
12749.0           3        199         4090.88           1           1   
12820.0           3         59          942.34           1           2   

            m_quartile RFMScore  
CustomerID                       
12346.0              1      441  
12747.0              1      111  
12748.0              1      111  
12749.0              1      111  
12820.0              2      122

In [40]:
segmented_rfm[segmented_rfm['RFMScore']=='111'].sort_values('monetary_value', ascending=False).head(10)

recency  frequency  monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
18102.0           0        431       259657.30           1           1   
17450.0           8        337       194550.79           1           1   
17511.0           2        963        91062.38           1           1   
16684.0           4        277        66653.56           1           1   
14096.0           4       5111        65164.79           1           1   
13694.0           3        568        65039.62           1           1   
15311.0           0       2379        60767.90           1           1   
13089.0           2       1818        58825.83           1           1   
15769.0           7        130        56252.72           1           1   
15061.0           3        403        54534.14           1           1   

            m_quartile RFMScore  
CustomerID                       
18102.0              1      111  
17450.0              1      111  
17511.0              1      111  
16684.0              1      111  
14096.0              1      111  
13694.0              1      111  
15311.0              1      111  
13089.0              1      111  
15769.0              1      111  
15061.0              1      111

These are top 10 Customers according to RFM

In [43]:
segmented_rfm[segmented_rfm['RFMScore']=='444'].sort_values('monetary_value', ascending=False).tail(10)

recency  frequency  monetary_value  r_quartile  f_quartile  \
CustomerID                                                               
15940.0         311          1           35.80           4           4   
14576.0         372          1           35.40           4           4   
16765.0         294          1           34.00           4           4   
17408.0         163          2           32.65           4           4   
13120.0         238          1           30.60           4           4   
17102.0         261          1           25.50           4           4   
15823.0         372          1           15.00           4           4   
17763.0         263          1           15.00           4           4   
17956.0         249          1           12.75           4           4   
16738.0         297          1            3.75           4           4   

            m_quartile RFMScore  
CustomerID                       
15940.0              4      444  
14576.0              4      444  
16765.0              4      444  
17408.0              4      444  
13120.0              4      444  
17102.0              4      444  
15823.0              4      444  
17763.0              4      444  
17956.0              4      444  
16738.0              4      444

These are worst 10 customers